text/0000-edition-based-method-disambiguation.md

```diff
@@ -0,0 +1,224 @@
+ambiguity.
+
+This flag should be usable to resolve the following forms of breakage:
+
+* A new method on an pre-edition trait ([e.g.
+  itertools::intersperse](https://github.com/rust-lang/rust/issues/88967))
+* A new trait implementation of an pre-edition trait on an pre-edition type ([e.g.
+  ErrorKind Display](https://github.com/rust-lang/rust/issues/94507))
+* A new inherent method on an pre-edition type (no recent examples)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Disambiguation can require invasive changes
+
+In simple cases where you only call a single method in an expression switching
+from the ambiguous method call syntax, `self.method(args...)` to the
+unambiguous function call syntax `Trait::method(self, args...)` is an easy
+change. In longer method call chains however there isn't a way to disambiguate
+the trait a method is associated with when calling it without splitting up that
+expression into multiple expressions, which can change drop behavior and
+prevent temporaries for living as long as they need to.
+
+This RFC intentionally avoids solving this problem or even proposing strawmen
+versions of the syntax to avoid distracting from the core issue, but at the
+same time it increases the need for a language syntax extension for quickly
+disambiguating the trait a method call should come from.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+[This
+comment](https://github.com/rust-lang/rust/issues/88967#issuecomment-938024847)
+on the `Iterator::intersperse` issue details a few alternatives that were
+discussed by the libs team when we encountered that specific issue. These
+include both short term and long term suggestions.
+
+**Short Term**
+
+* use build scripts in crates where this breakage is expected to detect when
+  the breakage is introduced by the compiler and disable the conflicting APIs
+  internally using `#[cfg]` directives.
+* use `#[cfg(accessible)]` in the crate where the breakage is expected to
+  automatically remove the conflicting API when the upstream API is introduced.
+
+These solutions don't solve the problem generally, and instead address the
+specific breakage that are known or expected. We can and do catch many such
+issues via crater but we cannot test all crates via crater and we still end up
+breaking people's crates on nightly before the crater runs have had a chance to
+catch any breakage.
+
+**Longer Term**
+
+* encoding the released std version in crates published to crates.io when
+  uploading and ignore methods introduced in newer releases when building that
+  published dependency.
+* [Supertrait item shadowing](https://github.com/rust-lang/rfcs/pull/2845)
+
+These solutions also do not solve the problem generally. The former solution
+would only solve the problem for crates uploaded to crates.io, and the latter
+would only solve the problem for traits where the ambiguity is introduced by a
+supertrait.
```

Comment:
I think the solution of using `rust-version` is dismissed a bit too quickly here. Using the `rust-version` property of `Cargo.toml` doesn't have to be tied to uploading to crates.io. In the absense of a `rust-version` attribute, we could fall back to the rust version in which the selected edition got stabilized, for example. (And never fall back to a version older than the one in which the version-dependent lookups started happening.)

Reply:
oh, I think I had misunderstood how this was intended to work. I'll update this section.

Reply:
My idea was to make it depend on `rust-version`, and also have `cargo publish` add that field on uploading if you didn't already have it. But if the second part turns out to be a bad idea, the first part can still work by itself. When combined with `edition` (such that e.g. `edition = "2021"` simply implies `rust-version = "1.56.0"`), I think it could work pretty well.

Reply:
I think this is worth considering, yes. Certainly rust-version could be a way to "opt-in" to newer methods *before* an edition is standardized, much like "from __future__ import" in Python?

Reply:
@nikomatsakis I mentioned this above in https://github.com/rust-lang/rfcs/pull/3240#issuecomment-1063086491 but neither of these approaches would gate new methods on edition releases.